fe/fe-core/src/main/java/com/starrocks/sql/optimizer/Optimizer.java

```diff
@@ -183,8 +195,74 @@ private void prepare(ConnectContext connectContext, OptExpression logicOperatorT
         context.setTraceInfo(new OptimizerTraceInfo(connectContext.getQueryId()));
 
         // process materialized views
-        if (Config.enable_experimental_mv) {
+        // TODO: add session variable
+        if (Config.enable_experimental_mv && !optimizerConfig.isRuleBased()) {
             // TODO: register materialized views
+            // register materialized views
+            registerMaterializedViews(logicOperatorTree, connectContext);
+        }
+    }
+
+    private void registerMaterializedViews(OptExpression logicOperatorTree, ConnectContext connectContext) {
+        List<Table> tables = Utils.getAllTables(logicOperatorTree);
+
+        // include nested materialized views
+        Set<MaterializedView> relatedMvs = Sets.newHashSet();
+        Utils.getRelatedMvs(tables, relatedMvs);
+
+        for (MaterializedView mv : relatedMvs) {
+            if (!mv.isActive()) {
+                continue;
+            }
+            Set<String> partitionNamesToRefresh = Sets.newHashSet();
+            // Set<String> partitionNamesToRefresh = mv.getPartitionNamesToRefreshForMv();
+            PartitionInfo partitionInfo = mv.getPartitionInfo();
+            if (partitionInfo instanceof SinglePartitionInfo) {
+                if (!partitionNamesToRefresh.isEmpty()) {
+                    continue;
+                }
+            } else if (partitionNamesToRefresh.containsAll(mv.getPartitionNames())) {
+                // if the mv is partitioned, and all partitions need refresh,
+                // then it can not be an candidate
+                continue;
+            }
+
+            // 1. build mv query logical plan
+            String mvSql = mv.getViewDefineSql();
+            ColumnRefFactory columnRefFactory = new ColumnRefFactory();
+            Pair<OptExpression, LogicalPlan> plans = Utils.getRuleOptimizedLogicalPlan(mvSql, columnRefFactory, connectContext);
```

Comment:
Please consider cache it in memory, otherwise it's really expensive

Reply:
here I do not cache it in memory because:
1. maybe some SessionVariable will change the plan generated
2. during each optimization, I will modify the mv plan to add some predicate to plan